src/tunnel.cc

```diff
@@ -947,8 +946,8 @@ tunnelConnectedWriteDone(const Comm::ConnectionPointer &conn, char *, size_t len
     tunnelState->client.writer = NULL;
 
     if (flag != Comm::OK) {
-        *tunnelState->status_ptr = Http::scInternalServerError;
-        tunnelErrorComplete(conn->fd, data, 0);
+        auto err = new ErrorState(ERR_WRITE_ERROR, Http::scInternalServerError, tunnelState->request.getRaw(), tunnelState->al);
+        tunnelState->sendErrorAndDestroy(err, "client write failure");
```

Comment:
Actually, since client.dirty must be true here to 'error sending' is possible. I think we should still call sendErrorAndDestroy() since it does some general cleanup steps.

Reply:
I agree. If possible, we should call sendErrorAndDestroy() on all errors and let that code deal with all the complex decisions, dependencies, cleanup, etc.

Reply:
There is one case in tunnel.cc where tunnelStart() where errorSend() is called. This case seems problematic because there is no 'client expects response' check there. However, we cannot call sendErrorAndDestroy() instead because the TunnelStateData object does not exist yet. Probably we need to set errorSend() conditioned on a new (static?) method doing what TunnelStateData::clientExpectsConnectResponse() does now.

Reply:
> Probably we need to set errorSend() conditioned on a new (static?) method doing what TunnelStateData::clientExpectsConnectResponse() does now

I would create TunnelStateData unconditionally and then, in a new TunnelStateData::start() method, do the Config.accessList.miss check. That way, you can use sendErrorAndDestroy() as usual. The overhead of creating TunnelStateData is probably relatively small, and Config.accessList.miss denials ought to be rare.

One could argue that some errors could happen even before ClientHttpRequest calls tunnelStart(). Who will apply the right error response (not) sending logic then? I think that is a valid concern, but I would leave that investigation outside this PR scope and focus on what happens _after_ the decision to tunnel has been made (i.e. tunnelStart() has been called).

Reply:
> Alex: I would create TunnelStateData unconditionally and then, in a new TunnelStateData::start() method, do the Config.accessList.miss check.

No, that would be wrong. The decision to tunnel is not made when we call tunnelStart(). It is made _inside_ tunnelStart(), after the Config.accessList.miss check. We must not use TunnelStateData logic before that check because doing so might (sooner or later) result in, for example, pre-buffered from-client data to be sent to the origin server, in violation of the failed Config.accessList.miss check.

Reply:
Most likely, the meat of TunnelStateData::clientExpectsConnectResponse() should become a new ClientHttpRequest::clientExpectsConnectResponse() method. The remaining TunnelStateData code would just check whether `http` is still valid (and return false if it is not). AFAICT, ClientHttpRequest and TunnelStateData share the same `request` object.

This will change the resulting code slightly, but I think that is OK: The `http` pointer is valid when TunnelStateData starts. Later on, I assume, if `http` becomes invalid (if that is possible at all), then the underlying client connection is being closed and, hence, the client is not really expecting a response anymore even if it were expecting it earlier so returning false is appropriate.

Reply:
> should become a new ClientHttpRequest::clientExpectsConnectResponse() method

Done at 9cfe98f and adjusted tunnelStart() accordingly.